Site Tagline Block

Copons opened the pull request:
## Description

Replaces https://github.com/WordPress/gutenberg/pull/18241

Add a Site Tagline (aka Site Description) block for the Site Editor.
It supports colors, font size, and line height out of the box.

## How has this been tested?

Tested in the Site Editor, on both Firefox 78 and Chrome 83 on macOS 10.15.5.

## Screenshots

<img width="659" alt="Screenshot 2020-07-13 at 18 31 46" src="https://user-images.githubusercontent.com/2070010/87334801-21d98c80-c537-11ea-9820-71967dc8a0bb.png">

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

Copons committed:
Site Description block initial commit

Copons committed:
Properly use the light block wrapper

Copons committed:
Add fixtures

ockham commented on packages/block-library/src/site-description/block.json:
```diff
@@ -0,0 +1,13 @@
+	"name": "core/site-description",
+	"category": "design",
+	"attributes": {
+		"align": {
+			"type": "string"
+		}
```
I _think_ you don't need to add that as an attribute but can just add it to `supports`. [See](https://developer.wordpress.org/block-editor/developers/block-api/block-registration/#supports-optional) for syntax.

Copons replied:
See https://github.com/WordPress/gutenberg/pull/23788#discussion_r452275107

ockham commented on packages/block-library/src/site-description/edit.js:
```diff
@@ -0,0 +1,49 @@
+	return (
+		<>
+			<BlockControls>
+				<AlignmentToolbar
+					onChange={ ( newAlign ) =>
+						setAttributes( { align: newAlign } )
+					}
+					value={ align }
+				/>
```
I think [adding `align` to `suppports`](https://github.com/WordPress/gutenberg/pull/23788/files#r452192964) will give you the `AlignmentToolbar` out f the box :crossed_fingers:

Copons replied:
@ockham The difference is that `align` in `supports` is the _block_ alignment, while in this case it would be the _text_ alignment.
Two major differences, off the top of my head:
- Block alignment also (optionally) supports `wide` and `full`.
- Block alignment `left` and `right` makes the block floating.

This said, your point makes sense in that we might want this block to support both block (as `align` `supports`) and text alignment (as `textAlign` attribute).
I don't see any other Core block doing the same thing, and I wonder if it's an intentional UX choice (it might be confusing for the user).
A middle ground could be to have `align` only support `wide, full`. 🤔

ockham replied:
Right, thanks for clarifying.

I do think that this has the potential of being quite confusing to the user, so I wouldn't be surprised if it was a deliberate UX choice.

Do blocks that are somewhat comparable to Site Description tend to have either align or text align?

If we do go with text align, let's rename the attribute to `textAlign`, to make it easier to distinguish from block align.

Copons replied:
The most comparable is arguably [Site Title](https://github.com/WordPress/gutenberg/blob/5dbff73802fe5c50bf0489dfee7d9bb7f44589a7/packages/block-library/src/site-title/edit.js), which indeed only has the text alignment with the `align` attribute name.

(It's not a coincidence, I've mostly copied it to build the Site Description block... 😄)

Copons replied:
> If we do go with text align, let's rename the attribute to textAlign, to make it easier to distinguish from block align.

I agree, but I'd love more Gutenberg folks chime in on this.
As far as I can see, only Verse uses `textAlign` for text alignment, while all other blocks simply use `align`.

Copons committed:
Add support for experimental colors, font size, line height

Copons commented:
Added support for experimental colors, font size, and line height (introduced in #23007).

Basically the Site Description block is almost identical to Site Title.
Differences:
- They use different site options (obviously).
- Site Description only outputs a paragraph element, whereas Site Title allows for paragraph or heading (1-6).

About the output element: I'm wondering if we might even "downgrade" it to a simple `div` instead of `p`, or give the choice.

Either way, I've chosen not to include the Site Title's level selector (to set it as heading element) following the path laid out by the Twenty themes: the last one using an `h2` for the site description was Twenty Fourteen; since then it has always been either `p` or `div`.

Copons committed:
Update name to Site Tagline

ockham commented:
> The icon is `info`, but of course I'm open to different suggestions.
For reference, the previous Site Description PR used [this custom icon](https://github.com/WordPress/gutenberg/blob/f65e5dd57718da6a90629a3dfe5d0257896d6e17/packages/block-library/src/site-description/icon.js):
> <img width="88" alt="Screenshot 2020-07-08 at 16 19 53" src="https://user-images.githubusercontent.com/2070010/86937231-e9047680-c136-11ea-8d7d-55fbbfe1758b.png">

I kinda like that, and I think it's less generic than the `info` icon. Let's use it maybe?

ockham reviewed (approved):
LGTM :+1:

Copons committed:
Add icon and keywords

Copons commented:
Thanks @ockham! I've updated the icon to use the one from the old PR.

I've also added a `description` keyword to the block definition.
While the block is now properly called "tagline" (user-facing label for the setting), I'd argue that WP developers are used to call it "description" (setting name used behind the scenes) instead — or at least _I_ am! 😄

jeyip reviewed (commented):
I spent a lot of time learning about `block.json` and `experimental-themes.json`. Not an expert with the Gutenberg metadata interface, but as far as the code goes, everything looks reasonable.

I did notice one curious behavior that affects all blocks that can modify colors (including `site-tagline`). To clarify, this seems best addressed in a follow-up PR.

For any block with a background color set as a direct child of a group block, text padding isn't appropriately applied. After some digging, it looks like padding styles are being overwritten by `group` block css [here](https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/group/editor.scss#L42-L71) on L51-52 and L59-60.

<img width="1824" alt="Screen Shot 2020-07-13 at 5 08 43 PM" src="https://user-images.githubusercontent.com/5414230/87365752-e7b6bd80-c52b-11ea-9bd4-e508139b22d3.png">
<img width="1825" alt="Screen Shot 2020-07-13 at 4 51 09 PM" src="https://user-images.githubusercontent.com/5414230/87365755-e8e7ea80-c52b-11ea-8cd2-9d75e1e1967f.png">


This occurs for me in chrome, firefox, and safari.

I'll make a separate issue for this in Gutenberg if that sounds good to you @Copons

Copons commented:
> I'll make a separate issue for this in Gutenberg if that sounds good to you @Copons

Sure @jeyip, go for it!
I've always had trouble figuring out the spacing of nested blocks.

jeyip reviewed (commented):
> I'll make a separate issue for this in Gutenberg if that sounds good to you @Copons

https://github.com/WordPress/gutenberg/issues/23937

Do you happen to know where we place these kinds of lower priority site editor specific bug tickets, Jacopo?

Copons commented:
> Do you happen to know where we place these kinds of lower priority site editor specific bug tickets, Jacopo?

@jeyip I'm not sure, let's ask @vindl! 🙂 
Meanwhile I've labeled it as FSE and bug, just in case.

One thing about this kind of visual issues that I didn't think of mentioning earlier (I haven't checked the issue yet, btw): they might be caused by the theme's own editor style, in which case Gutenberg might not be at fault.

If this is the case, we have a few options. For example:
- Check if the theme is doing it right, but Gutenberg style "gets in the way". We might want to improve Gutenberg style to be tighter, or less specific.
- Otherwise, the theme style is incorrect, and we should open an issue in the theme repo.

For experience, styling nested blocks in the editor is a major pain, especially when we go out of our way to customize the blocks margins.
It's just incredibly hard to test for all possible combinations, and I've never found a good enough solution that satisfies all requirements.

strarsis commented:
So this feature is planned to be released with a near future Gutenberg release?

noahtallen commented:
(see my response on #24246)
